README.md
# Seml

Implement a **monadic parser combinator** from scratch to parse S-Expression texts and transform it into XML.

### For anyone who may be interested in how to do such things
See `Src/Parser.hs` if you are interested in how to write a parser combinator.

See `Src/Syntax.hs` for how to use a parser combinator.

### Output Example
Original S-Expression:
```Lisp
(?xml (:version '1.0') (:encoding 'utf-8'))
(ksiazka-telefoniczna (:category 'bohaterowie książek')
  (osoba (:charakter 'dobry')
    (imie Ambroży)
    (nazwisko Kleks)
    (telefon 123-456-789))
  (osoba (:charakter 'zły')
    (imie Alojzy)
    (nazwisko Bąbel)
    (telefon)))
```

Parsed S-Expression:
```
[ Elem "?xml"
    [ Rec
        ( Elem ":version" [ Text "1.0" ] ), Rec
        ( Elem ":encoding" [ Text "utf-8" ] ) ], 
  Elem "ksiazka-telefoniczna"
    [ Rec
        ( Elem ":category" [ Text "bohaterowie książek" ] ), Rec
        ( Elem "osoba"
            [ Rec
                ( Elem ":charakter" [ Text "dobry" ] ), Rec
                ( Elem "imie" [ Text "Ambroży" ] ), Rec
                ( Elem "nazwisko" [ Text "Kleks" ] ), Rec
                ( Elem "telefon" [ Text "123-456-789" ] ) ] ), Rec
        ( Elem "osoba"
            [ Rec
                ( Elem ":charakter" [ Text "zły" ] ), Rec
                ( Elem "imie" [ Text "Alojzy" ] ), Rec
                ( Elem "nazwisko" [ Text "Bąbel" ] ), Rec
                ( Tag "telefon" ) ] ) ] ]
```

XML output:
```XML
<?xml version="1.0" encoding="utf-8" ?>
<ksiazka-telefoniczna category="bohaterowie książek">
  <osoba charakter="dobry">
    <imie>Ambroży</imie>
    <nazwisko>Kleks</nazwisko>
    <telefon>123-456-789</telefon>
  </osoba>
  <osoba charakter="zły">
    <imie>Alojzy</imie>
    <nazwisko>Bąbel</nazwisko>
    <telefon />
  </osoba>
</ksiazka-telefoniczna>
```


CHANGELOG.md
# Changelog for `Seml`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


test/Spec.hs
import Lib
import Test.HUnit

test1 :: Test
test1 =
    TestCase $
        let (_, val) = runParser semlP "(a\\ c)"
         in assertEqual "Seml1" (Right [Tag "a c"]) val

test2 :: Test
test2 =
    TestCase $
        let (_, val) = runParser semlP "(a ('xx\\''))"
         in assertEqual "Seml2" (Right [Elem "a" [Rec (Tag "xx'")]]) val

test3 :: Test
test3 =
    TestCase $
        let (_, val) = runParser semlP "(a b c)"
         in assertEqual "Seml3" (Right [Elem "a" [Text "b", Text "c"]]) val

main :: IO Counts
main =
    runTestTT . TestList $
        [test1, test2, test3]


src/Seml.hs
{-# LANGUAGE ImportQualifiedPost #-}

module Seml (Seml (..), Element (..), toXML) where

import Text.XML.Light (Attr, node, unqual)
import Text.XML.Light qualified as XML

data Seml = Elem String [Seml.Element] | Tag String
    deriving (Show, Eq)
data Element = Rec Seml | Text String
    deriving (Show, Eq)

extractAttr :: [Seml.Element] -> ([Attr], [Seml.Element])
extractAttr [] = ([], [])
extractAttr (x : xs) =
    let (attrs, elems) = extractAttr xs
     in case x of
            Rec (Elem (':' : name) [Text val]) -> (XML.Attr (unqual name) val : attrs, elems)
            Text s -> case elems of
                (Text s1) : rest -> (attrs, Text (s ++ s1) : rest)
                rest -> (attrs, Text s : rest)
            _ -> (attrs, x : elems)

toElem :: Element -> XML.Element
toElem (Text s) = node (unqual s) ()
toElem (Rec (Elem tag [Text text])) = node (unqual tag) ([] :: [Attr], text)
toElem (Rec s) = toXML s

toXML :: Seml -> XML.Element
toXML (Tag t) = node (unqual t) ()
toXML (Elem tag elems) =
    let (attr, elems') = extractAttr elems
        name = unqual tag
        xmlElems = map toElem elems'
     in node name (attr, xmlElems)


src/Parser.hs
{-# LANGUAGE TupleSections #-}

module Parser (Parser (Parser), runParser, (<|>), repeatP, betweenP, sepByP) where

import Control.Applicative (Applicative (liftA2))
import Control.Monad (ap)

-- | A very simple parser combinator
newtype Parser a = Parser {runParser :: String -> (String, Either String a)}

instance Functor Parser where
    fmap f p = Parser $ \s ->
        let (s', a) = runParser p s
         in (s', f <$> a)

instance Applicative Parser where
    pure x = Parser (,Right x)
    (<*>) = ap

instance Monad Parser where
    return = pure
    x >>= f = Parser $ \s ->
        let (s', a) = runParser x s
         in case a of
                Left err -> (s, Left err)
                Right val -> runParser (f val) s'

-- | Combine parser with an alternative one
(<|>) :: Parser a -> Parser a -> Parser a
p1 <|> p2 = Parser $ \s ->
    case runParser p1 s of
        (_, Left _) -> runParser p2 s
        a -> a

someP :: Parser a -> Parser [a]
someP p = liftA2 (:) p (manyP p)

manyP :: Parser a -> Parser [a]
manyP p = terminateP <|> someP p <|> return []

terminateP :: Parser [a]
terminateP = Parser $ \s ->
    case s of
        [] -> ([], Right [])
        _ -> (s, Left "")

-- | Repeat a parser
repeatP :: Parser a -> Parser [a]
repeatP = manyP

-- | Parse something between other things
betweenP :: Parser l -> Parser r -> Parser mid -> Parser mid
betweenP l r m = l *> m <* r

-- | Repeat parser with separator
sepByP :: Parser s -> Parser a -> Parser [a]
sepByP ps pa = liftA2 (:) pa (repeatP (ps *> pa))


src/Syntax.hs
module Syntax (semlP) where

import Control.Applicative (Applicative (liftA2))
import Data.Char (isSpace)
import Parser (Parser (Parser), betweenP, repeatP, sepByP, (<|>))
import Seml (Element (Rec, Text), Seml (Elem, Tag))

charP :: (Char -> Bool) -> Parser Char
charP f = Parser $ \s ->
    case s of
        (x : xs)
            | f x -> (xs, Right x)
            | otherwise -> (s, Left ("Invalid Char '" ++ x : ['\'']))
        [] -> (s, Left "Eof Error")

escapeSP :: Parser Char
escapeSP = isP '\\' *> charP (const True)

-- | Parse white spaces
spaceP :: Parser String
spaceP = repeatP (charP isSpace)

-- | Parse word with escaped white space
wordP :: Parser String
wordP =
    let wcP = escapeSP <|> charP (not . \x -> isSpace x || x == ')' || x == '(')
     in liftA2 (:) wcP (repeatP wcP)

isP :: Char -> Parser Char
isP c = charP (== c)

-- | Parse quoted string
quotedP :: Parser String
quotedP =
    let ssP = repeatP (escapeSP <|> charP (/= '\''))
        dsP = repeatP (escapeSP <|> charP (/= '"'))
        singleP = betweenP (isP '\'') (isP '\'') ssP
        doubleP = betweenP (isP '"') (isP '"') dsP
     in singleP <|> doubleP

semlOneP :: Parser Seml
semlOneP =
    let tagP' = Tag <$> tagP
        elemsP' = liftA2 Elem (tagP <* spaceP) elemsP
        betweenPP = betweenP (isP '(' <* spaceP) (spaceP *> isP ')')
     in betweenPP tagP' <|> betweenPP elemsP'

-- | Seml Parser
semlP :: Parser [Seml]
semlP = repeatP $ spaceP *> semlOneP <* spaceP

tagP :: Parser String
tagP = quotedP <|> wordP

elemP :: Parser Element
elemP =
    let elemWordP = Text <$> wordP
        elemStringP = Text <$> quotedP
        elemRecP = Rec <$> semlOneP
     in elemStringP <|> elemWordP <|> elemRecP

elemsP :: Parser [Element]
elemsP = sepByP spaceP elemP


app/Main.hs
module Main (main) where

import Control.Monad.List (forM_)
import Lib
import Text.Pretty.Simple (CheckColorTty (CheckColorTty), OutputOptions (outputOptionsCompact, outputOptionsCompactParens), defaultOutputOptionsLightBg, pPrintOpt)
import Text.XML.Light (ppElement)

text :: String
text =
    "(?xml (:version '1.0') (:encoding 'utf-8'))\n\
    \(ksiazka-telefoniczna (:category 'bohaterowie książek')\n\
    \  (osoba (:charakter 'dobry')\n\
    \    (imie Ambroży)\n\
    \    (nazwisko Kleks)\n\
    \    (telefon 123-456-789))\n\
    \  (osoba (:charakter 'zły')\n\
    \    (imie Alojzy)\n\
    \    (nazwisko Bąbel)\n\
    \    (telefon)))"

parse :: String -> Either String [Seml]
parse = snd . runParser semlP

printOutput :: String -> IO ()
printOutput inp = do
    case parse inp of
        Left err -> putStrLn $ "Error : " ++ err
        Right val -> do
            putStrLn "Parsed S-Expression:"
            let opts = defaultOutputOptionsLightBg{outputOptionsCompact = True, outputOptionsCompactParens = True}
             in pPrintOpt CheckColorTty opts val >> putStrLn ""
            putStrLn "XML output:"
            forM_ (map (ppElement . toXML) val) putStrLn

main :: IO ()
main = do
    x <- getContents
    putStrLn "Enter original S-Expression(EOF to finish/left empty will display example):"
    inp <- case x of
        [] -> putStrLn text >> return text
        inp -> putStrLn "" >> return inp
    printOutput inp
